tests/test_date.py

```diff
@@ -132,76 +205,180 @@ def test_periods(self, low, high, period, expected_results):
+              mock_date=datetime(2014, 8, 12),
+              date_str='August 2014',
+              date_format='%B %Y'),
+        param(given_date=datetime(2014, 2, 28),
+              mock_date=datetime(2014, 2, 12),
+              date_str='February 2014',
+              date_format='%B %Y'),
+    ])
+    def test_should_use_current_date_for_dates_without_day(self, given_date, mock_date,
+                                                           date_str, date_format):
+        self.given_date(given_date)
+        self.given_datetime_mock(mock_date)
+        self.given_patch()
+        self.when_patch_parsed(date_str, date_format)
+        self.then_dates_is_not_none()
+        self.then_period_equal()
+        self.then_dates_are_equal()
+
+    def given_invalid_date(self, date_str):
+        self.invalid_date = date_str
+
+    def given_current_year(self):
+        self.year = datetime.today().year
+
+    def given_date_format(self, date_format):
+        self.format = date_format
+
+    def given_date(self, date_obj):
+        self.date_obj = date_obj.date()
+
+    def given_date_str(self, date_str):
+        self.date_str = date_str
 
-        result = date.parse_with_formats('09.16', ["%m.%d"])
-        self.assertEquals(datetime(today.year, 9, 16).date(), result['date_obj'].date())
+    def given_datetime_mock(self, given_date):
+        self.datetime_mock = Mock(wraps=datetime)
+        self.datetime_mock.utcnow = Mock(return_value=given_date)
 
-    def test_should_use_current_date_for_dates_without_day(self):
-        twelfth = datetime(2014, 8, 12)
-        datetime_mock = Mock(wraps=datetime)
-        datetime_mock.utcnow = Mock(return_value=twelfth)
+    def given_patch(self):
+        self.add_patch(patch('dateparser.date_parser.datetime', new=self.datetime_mock))
 
-        with patch('dateparser.date_parser.datetime', new=datetime_mock):
-            dt_data = date.parse_with_formats('August 2014', ['%B %Y'])
+    def when_should_use_current_year_for_dates_without_year(self):
+        self.result = date.parse_with_formats('09.16', ["%m.%d"])
 
-        self.assertIsNotNone(dt_data)
-        self.assertEquals('month', dt_data['period'])
-        self.assertEquals(datetime(2014, 8, 31).date(), dt_data['date_obj'].date())
+    def when_parse_invalid_date(self):
+        self.parsed_invalid_date = date.parse_with_formats(self.invalid_date,
+                                                           ['%Y-%m-%d'])['date_obj']
 
-        twelfth = datetime(2014, 2, 12)
-        datetime_mock = Mock(wraps=datetime)
-        datetime_mock.utcnow = Mock(return_value=twelfth)
+    def when_should_parse_date(self):
+        self.result = date.parse_with_formats(self.date_str, [self.format])
 
-        with patch('dateparser.date_parser.datetime', new=datetime_mock):
-            dt_data = date.parse_with_formats('February 2014', ['%B %Y'])
+    def when_patch_parsed(self, some_date, some_format):
```

Comment:
All four `when_*` methods here are really following one pattern: `result = date.parse_with_formats(date_string, [date_format])`. So let's use `given_date_str` and `given_date_format` in test and refactor all four methods to `when_date_is_parsed_with_format`.
Take a look if there are other cases alike in this pull-request.